### ISSUE TITLE
Different behaviour between train and validation loader.

### ISSUE THREAD

**Issue by fgdfgfthgr-fox** [janelia-cellmap/cellmap-segmentation-challenge#163]
ecs class, 32nm resolution, 256x256x256 size. Using the included get_dataloader() class.
Training batches looks like this:
![Image](https://github.com/user-attachments/assets/f4909bde-912e-41f8-a52e-a1d4417b36f7)
Validation batches looks like this:
![Image](https://github.com/user-attachments/assets/bb8bc51e-b2af-4e58-971b-a74dc11d826e)

Regenerate datasplit does not help.

**Comment by rhoadesScholar** [comment]
Can you supply a minimal script that produces the behavior?

**Comment by fgdfgfthgr-fox** [comment]
> Can you supply a minimal script that produces the behavior?
```
from cellmap_segmentation_challenge.utils import get_dataloader
from cellmap_data.utils import get_fig_dict

if __name__ == "__main__":
    request_array_info = {
        "shape": (256, 256, 256),
        "scale": (32, 32, 32),
    }

    train_loader, val_loader = get_dataloader('datasplit.csv', ["ecs"], 8,
                                              input_array_info=request_array_info, target_array_info=request_array_info,
                                              device='cpu',
                                              iterations_per_epoch=1024,
                                              use_mutual_exclusion=True,
                                              num_workers=0, pin_memory=True, persistent_workers=False)

    for i in range(1):
        val_loader_i = iter(val_loader.loader)
        pair = next(val_loader_i)
        inputs = pair['input']
        targets = pair['output']
        outputs = pair['output']
        figs = get_fig_dict(inputs, targets, outputs, ["ecs", ])
        figs['ecs'].show()
    for i in range(1):
        train_loader_i = iter(train_loader.loader)
        pair = next(train_loader_i)
        inputs = pair['input']
        targets = pair['output']
        outputs = pair['output']
        figs = get_fig_dict(inputs, targets, outputs, ["ecs", ])
        figs['ecs'].show()
```

**Comment by rhoadesScholar** [comment]
@fgdfgfthgr-fox Ah. Could you include the `datasplit.csv` too, or the command(s) you used to construct it?

**Comment by fgdfgfthgr-fox** [comment]
> [@fgdfgfthgr-fox](https://github.com/fgdfgfthgr-fox) Ah. Could you include the `datasplit.csv` too, or the command(s) you used to construct it?

I already said the datasplit csv doesn't matter since different split produces similar result. Nevertheless, here it is.

[datasplit.csv](https://github.com/user-attachments/files/20402486/datasplit.csv)

**Comment by rhoadesScholar** [comment]
> > [@fgdfgfthgr-fox](https://github.com/fgdfgfthgr-fox) Ah. Could you include the `datasplit.csv` too, or the command(s) you used to construct it?
> 
> I already said the datasplit csv doesn't matter since different split produces similar result. Nevertheless, here it is.
> 
> [datasplit.csv](https://github.com/user-attachments/files/20402486/datasplit.csv)

Actually, the issue is largely related to the datasplit - that and whether or not you randomly shuffle the validation blocks you load. 

Try this datasplit (which flips the training vs. validation crops from the one you posted): [datasplit_flipped.csv](https://github.com/user-attachments/files/20593773/datasplit_flipped.csv)

Then run the following script that uses `random_validation=True` to shuffle the data loading for validation, like in training:
```
# %%
from cellmap_segmentation_challenge.utils import (
    get_dataloader,
    make_datasplit_csv,
)
from cellmap_data.utils import get_fig_dict
import numpy as np

scale = 32
shape = 256
classes = ["ecs"]
batch_size = 8
random_seed = 42


request_array_info = {
    "shape": (shape,) * 3,
    "scale": (scale,) * 3,
}

np.random.seed(random_seed) # For reproducibility

# %%
train_loader, val_loader = get_dataloader(
    # "datasplit.csv",
    "datasplit_flipped.csv",
    classes,
    batch_size=batch_size,
    input_array_info=request_array_info,
    target_array_info=request_array_info,
    device="cpu",
    iterations_per_epoch=1024,
    use_mutual_exclusion=True,
    num_workers=0,
    pin_memory=True,
    persistent_workers=False,
    random_validation=True,  # <--------
)

# %% VALIDATION LOADER
pair = next(iter(val_loader.loader))
inputs = pair["input"]
targets = pair["output"]
outputs = pair["output"]
figs = get_fig_dict(
    inputs,
    targets,
    outputs,
    classes,
)
figs[classes[0]].show()
figs[classes[0]].suptitle("Validation")
figs[classes[0]].tight_layout()

# %% TRAINING LOADER
pair = next(iter(train_loader.loader))
inputs = pair["input"]
targets = pair["output"]
outputs = pair["output"]
figs = get_fig_dict(
    inputs,
    targets,
    outputs,
    classes,
)
figs[classes[0]].show()
figs[classes[0]].suptitle("Training")
figs[classes[0]].tight_layout()

```

**Comment by rhoadesScholar** [comment]
Ah. My apologies. Please update cellmap-data again (i.e. `pip install --upgrade cellmap-data`).

**Comment by fgdfgfthgr-fox** [comment]
Updated again resolves the issue, but these very strange labels still exist. (note these are taken at 32nm with 128x128x128)

![Image](https://github.com/user-attachments/assets/b396c2d8-e63e-4e67-840c-7470d777f489)

![Image](https://github.com/user-attachments/assets/6b13710f-68b7-447d-bd7a-d19391c87b5b)

**Comment by fgdfgfthgr-fox** [comment]
Also, for a dataset that's 40+GB in size, I would expect more than 126 possible samples (especially in this setting, even 1px shift is considered a new sample) for a common class as extracellular space at 32nm... Right?

**Comment by rhoadesScholar** [comment]
> The neuroglancer doesn't really work since the script quit immediately after the url has been displayed.

@fgdfgfthgr-fox You must run the script in an interactive python instance or Jupyter notebook, or add this to the end to keep it from exiting immediately:
```
while True:
    pass
```

**Comment by rhoadesScholar** [comment]
@fgdfgfthgr-fox If you once again update cellmap-data, I've added a utility that will open the public link to the s3 data, based on the metadata of a batch:
```
from cellmap_data.utils import get_neuroglancer_link
print(get_neuroglancer_link({'raw_path': '/mnt/2878EBCCAED823C6/PycharmProjects/cellmap/cellmap-segmentation-challenge/data/jrc_ctl-id8-1/jrc_ctl-id8-1.zarr/recon-1/em/fibsem-uint8', 'current_center': {'z': 13597.419999999998, 'y': 5650.0, 'x': 32058.0}, 'current_idx': array(1), 'target_path_str': '/mnt/2878EBCCAED823C6/PycharmProjects/cellmap/cellmap-segmentation-challenge/data/jrc_ctl-id8-1/jrc_ctl-id8-1.zarr/recon-1/labels/groundtruth/crop117/{label}', 'class_weights': {'ecs': 1.4176668283607279}})
```

Simply copy and paste the link in your browser.

**Comment by fgdfgfthgr-fox** [comment]
> Could you be more specific regarding what you are seeing as errors in the segmentations?

Nevermind, now I think they are correct.
